metricbeat/module/prometheus/query/data.go

```diff
@@ -0,0 +1,161 @@
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package query
+
+import (
+	"encoding/json"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+)
+
+// ArrayResponse is for "scalar", "string" type.
+type ArrayResponse struct {
+	Status string    `json:"status"`
+	Data   arrayData `json:"data"`
+}
+type arrayData struct {
+	ResultType string        `json:"resultType"`
+	Results    []interface{} `json:"result"`
+}
+
+// MapResponse is for "vector", "matrix" type from Prometheus Query API Request
+type MapResponse struct {
+	Status string  `json:"status"`
+	Data   mapData `json:"data"`
+}
+type mapData struct {
+	ResultType string      `json:"resultType"`
+	Results    []mapResult `json:"result"`
+}
+type mapResult struct {
+	Metric  map[string]string `json:"metric"`
+	Vector  []interface{}     `json:"value"`
+	Vectors [][]interface{}   `json:"values"`
+}
+
+func parseResponse(body []byte, pathConfig QueryConfig) ([]mb.Event, error) {
+	var events []mb.Event
+	var resultType string
+	var convertedMap MapResponse
+
+	// try to convert to array response
+	convertedArray, err := convertJSONToArrayResponse(body)
+	if err != nil {
+		return events, err
+	}
+	resultType = convertedArray.Data.ResultType
+
+	// check if it is a vector or matrix and unmarshal more
+	if resultType == "vector" || resultType == "matrix" {
+		convertedMap, err = convertJSONToMapResponse(body)
+		if err != nil {
+			return events, err
+		}
+		resultType = convertedMap.Data.ResultType
+	}
+
+	switch resultType {
+	case "scalar", "string":
+		if convertedArray.Data.Results != nil {
+			events = append(events, mb.Event{
+				Timestamp: getTimestamp(convertedArray.Data.Results[0].(float64)),
+				MetricSetFields: common.MapStr{
+					"dataType":           resultType,
+					pathConfig.QueryName: attemptConvertToNumeric(convertedArray.Data.Results[1].(string)),
+				},
+			})
+		} else {
+			return events, errors.New("Could not retrieve results")
+		}
+	case "vector":
+		for _, result := range convertedMap.Data.Results {
+			if result.Vector != nil {
+				events = append(events, mb.Event{
+					Timestamp:    getTimestamp(result.Vector[0].(float64)),
+					ModuleFields: common.MapStr{"labels": result.Metric},
+					MetricSetFields: common.MapStr{
+						"dataType":           resultType,
+						pathConfig.QueryName: attemptConvertToNumeric(result.Vector[1].(string)),
```

Comment:
I'm not digging into the mechanics of this code very much but the logic of `attemptConvertToNumeric` looks a lot like a code smell. It can return an `int`, a `float64` or a `string`. I don't know if this is intended but if this ends in Elasticsearch it must be only one kind of `type`. This should be covered by the compiler but because it's returning an `interface{}` being set into a type `interface{}` again, it doesn't.

Reply:
I think we can fix this like this:
1. if it is "string" dataType then we can cast it into string
2. if it is "scalar" datatype then we can cast it into float
3. in case of instant and range vectors the result should always be numeric so we can cast into float type.